dali/kernels/imgproc/resample/resampling_impl.cuh

```diff
@@ -0,0 +1,280 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef DALI_KERNELS_IMGPROC_RESAMPLE_RESAMPLING_IMPL_CUH_
+#define DALI_KERNELS_IMGPROC_RESAMPLE_RESAMPLING_IMPL_CUH_
+
+#include <cuda_runtime.h>
+#include "dali/kernels/static_switch.h"
+#include "dali/kernels/common/convert.h"
+#include "dali/kernels/imgproc/resample/resampling_filters.cuh"
+
+namespace dali {
+namespace kernels {
+
+constexpr int ResampleSharedMemSize = 32<<10;
+
+namespace {
+
+namespace resample_shared {
+  extern __shared__ float coeffs[];
+};
+
+/// @brief Implements horizontal resampling for a custom ROI
+/// @param x0 - start column, in output coordinates
+/// @param x1 - end column (exclusive), in output coordinates
+/// @param y0 - start row
+/// @param y1 - end row (exclusive)
+/// @tparam static_channels - number of channels, if known at compile time
+///
+/// The function fills the output in block-sized vertical spans.
+/// Block horizontal size is warp-aligned.
+/// Filter coefficients are pre-calculated for each vertical span to avoid
+/// recalculating them for each row, and stored in a shared memory block.
+///
+/// The function follows different code paths for static and dynamic number of channels.
+/// For the dynamic, the innermost loop goes over filter taps, which eliminates the neeed
+/// for thread-local memory to store intermediate sums. This allows processing arbitrary
+/// number of channels.
+/// For static number of channels, the run-time parameter `channels` is ignored and
+/// there's also a local temporary storage for a tap sum for each channel. This is faster,
+/// but requires extra registers for the intermediate sums.
+template <int static_channels = -1, typename Dst, typename Src>
+__device__ void ResampleHorz_Channels(
+    int x0, int x1, int y0, int y1,
+    float src_x0, float scale,
+    Dst *out, int out_stride,
+    const Src *in, int in_stride, int in_w, int dynamic_channels,
+    ResamplingFilter filter, int support) {
+  using resample_shared::coeffs;
+
+  src_x0 += 0.5f * scale - 0.5f - filter.anchor;
+
+  const float filter_step = filter.scale;
+
+  const int coeff_base = threadIdx.x;
+  const int coeff_stride = blockDim.x;
+  const int channels = static_channels < 0 ? dynamic_channels : static_channels;
+
+  const float bias = std::is_integral<Dst>::value ? 0.5f : 0;
+
+  for (int j = x0; j < x1; j+=blockDim.x) {
+    int dx = j + threadIdx.x;
+    const float sx0f = dx * scale + src_x0;
+    const int sx0 = ceilf(sx0f);
+    float f = (sx0 - sx0f) * filter_step;
+    __syncthreads();
```

Comment:
Maybe it is not needed here?

Reply:
It's very necessary. I spent a few hours looking for the lack of it. It guards the shared filter against being overwritten while some warps still use the previous one.